Resumable Upload: Add guidance on upload creation strategy (#2910)
httpwg/http-extensions: draft-ietf-httpbis-resumable-upload.md

The server might process the uploaded data and make its results available in another resource during or after the upload. This subsequent resource is different from the upload resource created by the upload creation request ({{upload-creation}}). The subsequent resource does not handle the upload process itself, but instead facilitates further interaction with the uploaded data. The server MAY indicate the location of this subsequent resource by including the `Content-Location` header field in the informational or final responses generated while creating ({{upload-creation}}), appending to ({{upload-appending}}), or retrieving the offset ({{offset-retrieving}}) of an upload. For example, a subsequent resource could allow the client to fetch information extracted from the uploaded data.

# Upload Strategies

> The definition of the upload creation request ({{upload-creation}}) provides the client with flexibility to choose whether the file is fully or partially transferred in the first request, or if no file data is included at all. The decision which behavior is best largely depends on the client's capabilities, its intention to avoid data transmission, and its knowledge about the server's support for resumable uploads.

Reviewer: ```suggestion
The definition of the upload creation request ({{upload-creation}}) provides the client with flexibility to choose whether the file is fully or partially transferred in the first request, or if no file data is included at all. Which behavior is best largely depends on the client's capabilities, its intention to avoid data transmission, and its knowledge about the server's support for resumable uploads.
```

The following subsections describe three different, typical upload strategies that are suited for common environments. Note that these modes are never explicitly communicated to the server and clients are not required to stick to one strategy, but can mix and adapt them to their needs.

## Optimistic Upload Creation

> An "optimistic upload creation" can be used if the client knows or assumes with a high degree of confidence that the server supports resumable uploads under the target URI and that the client is capable of handling and processing interim responses. The upload creation request then includes the full file because the client assumes that the file will be transferred without interruptions.

Reviewer 2: >... because the client assumes that the file will be transferred without interruptions.

> ... because the client assumes that the file will be transferred without interruptions or resumed if an interruption occurs.

> The benefit of this method is that if the upload creation request succeeds, the file was transferred in a single request avoiding additional round trips. A drawback is that the client may be restricted in its ability to resume that upload. If the upload was interrupted before the client received a `104 (Upload Resumption Supported)` intermediate response with the upload URL, the client cannot resume that upload due to the missing upload URL. The intermediate response might not be received if the interruption happens too early in the message exchange, the server does not support sending the `104 (Upload Resumption Supported)` intermediate response, or an intermediary dropped the intermediate response. The client must either ultimately fail the entire upload or retry the upload creation request if this is allowed by the application at the expense of transferring file data again.

Reviewer: ```suggestion
The benefit of this method is that if the upload creation request succeeds, the file was transferred in a single request without additional round trips.

A possible drawback is that the client might be unable to resume an upload. If an upload is interrupted before the client received a `104 (Upload Resumption Supported)` intermediate response with the upload URL, the client cannot resume that upload due to the missing upload URL. The intermediate response might not be received if the interruption happens too early in the message exchange, the server does not support sending the 104 (Upload Resumption Supported) intermediate response, or an intermediary dropped the intermediate response. Without a 104 response, the client needs to either treat the upload as failed or retry the entire upload creation request if this is allowed by the application.
```

I'll note that if the 104 doesn't have time to arrive, then the client isn't really losing much progress.  As long as they are able to start over, this seems like a very reasonable approach.

> The "optimistic upload creation" attempts to upload the file in one request at the potential expense of either not finishing the upload at all or re-transmitting file data again.
> 

Reviewer: ```suggestion
```

This is redundant.

## Careful Upload Creation

With a "careful upload creation" the client sends an empty upload creation request without including any file data. Upon successful response reception, the client can use the included upload URL to transmit the file data ({{upload-appending}}) and resume the upload at any stage if an interruption occurs. The retransmission of file data or the ultimate upload failure that can happen with an "optimistic upload creation" is therefore avoided at the expense of an additional request that does not carry file data.

This approach best suited if the client cannot receive intermediate responses, e.g. due to a limitation in the provided HTTP interface, or if large files are transferred where the cost of the additional request is miniscule compared to the effort of transferring the large file itself.

> ## Upgrading To Resumable Uploads

Reviewer: This seems indistinguishable from the optimistic approach to me.  The use of Upload-Complete might distinguish it, but I don't see how these are functionally different.

Author: You're correct. I merged them.

The approach specified in this document allows clients and servers to automatically upgrade non-resumable uploads to resumable ones. In a non-resumable upload, the file is transferred in a single request, usually `POST` or `PUT`, without any ability to resume from interruptions. The client can offer the server to upgrade such an upload to a resumable upload. If supported by the server, the file transfer benefits from the gained ability to resume. Otherwise, the transfer falls back to a non-resumable upload without additional cost.

> To perform an upgrade, the client adds the `Upload-Complete: ?1` header field to the original request, which attempts to transmit the entire file in one request. A server that supports resumable uploads at the target URI, can create a resumable upload resource and send its upload URL in a `104 (Upload Resumption Supported)` intermediate response for the client to resume the upload after interruptions. On the other hand, if the server does not support resumable uploads or does not want to upgrade to a resumable upload for this request, it can ignore the `Upload-Complete: ?1` header and treat the request as a non-resumable upload, where the client cannot recover from interruptions.

Reviewer: ```suggestion
To perform an upgrade, the client adds the `Upload-Complete: ?1` header field to the original request, which attempts to transmit the entire file in one request. A server that supports resumable uploads at the target URI, can create a resumable upload resource and send its upload URL in a 104 (Upload Resumption Supported) intermediate response for the client to resume the upload after interruptions.

A server that does not support resumable uploads or does not want to upgrade to a resumable upload for this request can ignore the `Upload-Complete: ?1` header and treat the request as a non-resumable upload.
```

# Upload Strategies

The definition of the upload creation request ({{upload-creation}}) provides the client with flexibility to choose whether the file is fully or partially transferred in the first request, or if no file data is included at all. Which behavior is best largely depends on the client's capabilities, its intention to avoid data re-transmission, and its knowledge about the server's support for resumable uploads.

The following subsections describe two typical upload strategies that are suited for common environments. Note that these modes are never explicitly communicated to the server and clients are not required to stick to one strategy, but can mix and adapt them to their needs.

## Optimistic Upload Creation

> An "optimistic upload creation" can be used if the client assumes that the server supports resumable uploads under the target URI and that the client is capable of handling and processing interim responses. The upload creation request then includes the full file because the client anticipates that the file will be transferred without interruptions or resumed if an interruption occurs.

Reviewer 3: > assumes that the server supports resumable uploads under the target URI

This seems to be a bit in conflict with the "Upgrading To Resumable Uploads" subsection now that they are merged. Maybe we can say "assumes that the server supports resumable uploads under the target URI, or has no knowledge of the server support", but then that sentence doesn't really say anything.

"Careful Upload Creation" is where the server must support resumable uploads. "Optimistic Upload Creation" is OK if the server doesn't.

Author: Yes, the client's knowledge about the server support was a difference of the two sections at first. Now that they are merged, the text gets a bit more unwieldy, but a added https://github.com/httpwg/http-extensions/commit/de5992146b5957168dc2c7e0af2a2dbe64086ab5 to correct the statements about the client's knowledge. What do you think? Is that sufficient or do we need more significant changes?

Reviewer 3: Looks good, thanks
